File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    r=[]
    for i,c in enumerate(dartResult):
        if c in ['S','D','T']:
            if dartResult[i-2] in ['0','1','2','3']:
                n=int(dartResult[i-2:i])
            else:
                n=int(dartResult[i-1])
            r+=[{'S':n**1,'D':n**2,'T':n**3}.get(c)]
        if c=='#':
            r[-1]=-1*r[-1]
        elif c=='*':
            if len(r)==1:
                r[-1]=r[-1]*2
            else:
                r[-1]=r[-1]*2
                r[-2]=r[-2]*2
    return sum(r)

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.cpp
#include <string>
#include <regex>
#include <vector>
#include <numeric>
#include <cmath>

using namespace std;

int solution(string dartResult) {
    int r=0,num;
    smatch dart;
    vector<string> dartlist;
    vector<int> dartnum;
    regex pattern {"\\d+[SDT][*#]?"};
    while(regex_search(dartResult,dart,pattern)) {
        dartlist.push_back(dart.str());
        dartResult=dart.suffix();
    }
    for(int i=0;i<dartlist.size();i++) {
        if(dartlist[i].find("S")!=string::npos) {
            num=pow(stoi(dartlist[i].substr(0,dartlist[i].find("S"))),1);
        } else if(dartlist[i].find("D")!=string::npos) {
            num=pow(stoi(dartlist[i].substr(0,dartlist[i].find("D"))),2);
        } else if(dartlist[i].find("T")!=string::npos) {
            num=pow(stoi(dartlist[i].substr(0,dartlist[i].find("T"))),3);
        }
        if(dartlist[i].find("*")!=string::npos) {
            if(dartnum.size()>0)
                dartnum[i-1]*=2;
            num*=2;
        } else if(dartlist[i].find("#")!=string::npos) {
            num*=-1;
        }
        dartnum.push_back(num);
    }
    r=accumulate(dartnum.begin(),dartnum.end(),0);
    return r;
}